server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java

```diff
@@ -52,6 +55,7 @@
     private final BigArrays bigArrays;
     private final IntConsumer multiBucketConsumer;
     private IntArray docCounts;
+    private DocCountProvider fieldDocCountProvider = DocCountProvider.DEFAULT_PROVIDER;
```

Comment:
I wonder if we can have a single instance of this provider per request ? Maybe it should be added in the `SearchContext` and re-initiated every time we need to perform a collection ? With @nik9000 refactoring of the breadth first execution I think it should be easier to plug the logic on a higher level.